### PR TITLE
cloudml-template - fixes and added functionality

### PR DESCRIPTION
**Pull request by ksalama**
@nkashy1 @puneith @thedriftofwords: Hello, kindly review this PR to the cloudml-template, which includes the following modifications:

1. **Fixed the links in the Readme.md file**, pointing to the modules in the template (they were pointing to the old repo, which doesn't exist anymore).
2. **Added learning rate exponential decay functionality**, using [tf.train.exponential_decay](https://www.google.co.uk/search?q=tf+exponential+decay&oq=tf+exponen&aqs=chrome.0.0j69i57j69i60l3j0.8952j0j4&sourceid=chrome&ie=UTF-8). The decay factor is a user supplied param (in task.py) and defaults to 1.0 (which means no decay). This is described in the param help.
3. **Added defining custom eval metric to canned estimators**, using [tf.contrib.estimator.add_metric](tf.contrib.estimator.add_metric). By default, no  additional metrics are specified.

### CODE REVIEW

**Comment by ksalama** [comment]
@puneith , @thedriftofwords , @rhaertel80, @elibixby - 

I have added more commits to this PR, which included the following changes:
1. Updated the **create_custom_estimator** function in model.py to use the **"heads"** APIs in the tf.contrib.estimater.
2. Updated the babyweight-custom example to use **tf.contrib.regression_head** (in order to follow the updated template.
3. Added a new example, **german-custom**, to show how to build a custom estimator for classification using **tf.contrib.estimator.binary_classification_head**.
4. **Employing the pervious modifications** in the two "custom" examples (i.e., adding a new evaluation metric using tf.contrib.estimator.add_metric, and exponential learning rate decay).
5. Updating the **Readme.md** file to refer to the **newly added example** and its relevant **dataset**.

File: cloudml-template/examples/babyweight-custom/trainer/model.py [hunk 1]
```diff
@@ -147,65 +129,61 @@ def _model_fn(features, labels, mode):
...
-                                                    loss=loss,
-                                                    train_op=train_op,
-                                                    eval_metric_ops=eval_metric_ops,
-                                                    predictions=predictions,
-                                                    export_outputs=export_outputs
-                                                    )
-        return estimator_spec
+        return logits
+
+    def _train_op_fn(loss):
```
Comment (ispirmustafa) [inline cloudml-template/examples/babyweight-custom/trainer/model.py]: we typically define train_op_fn inside the model_fn. Reason: many times it needs to access tensors produces within the model_fn.

File: cloudml-template/README.md [hunk 2]
```diff
@@ -13,13 +13,22 @@ Although the template provides standard implementation to different functionalit
...
-
-2. **scripts**: The directory includes scripts to 1) train the model locally, 2) train the model on Cloud ML Engine, 
-and 3) deploy the model on GCP as well as to make prediction (inference) using the deployed model.
-
-3. **examples**: Currently three different examples are included, 1) classification, using the [Census Income](https://archive.ics.uci.edu/ml/datasets/Census+Income) (UCI Machine Learning Repository) dataset, 2) regression, using the [House Pricing](https://www.kaggle.com/apratim87/housingdata/data) (Kaggle) dataset,
-and 3) regression, with a custom estimator, using [Natality Baby Weight](https://catalog.data.gov/dataset?tags=birth-weight) (data.gov) dataset.
+1. **[template](https://github.com/GoogleCloudPlatform/cloudml-samples/tree/master/cloudml-template/template)**: The directory includes: 
+    1) trainer template with all the python modules to adapt to your data, 
+    2) setup.py, 
+    3) confg.yaml file for hyper-parameter tuning and specifying the Cloude MLE scale-tier, and
```
Comment (thedriftofwords) [inline cloudml-template/README.md]: "confg.yaml" -> "config.yaml", looks like a typo

"Cloude MLE" -> "Cloud ML Engine"

File: cloudml-template/README.md [hunk 3]
```diff
@@ -13,13 +13,22 @@ Although the template provides standard implementation to different functionalit
...
 
 ### Repository Structure
-1. **template**: The directory includes 1) trainer template with all the python modules to adapt to your data, 2) setup.py, and 3) confg.yaml file for hyper-parameter tuning and specifying the Cloude MLE scale-tier.
-
-2. **scripts**: The directory includes scripts to 1) train the model locally, 2) train the model on Cloud ML Engine, 
-and 3) deploy the model on GCP as well as to make prediction (inference) using the deployed model.
-
-3. **examples**: Currently three different examples are included, 1) classification, using the [Census Income](https://archive.ics.uci.edu/ml/datasets/Census+Income) (UCI Machine Learning Repository) dataset, 2) regression, using the [House Pricing](https://www.kaggle.com/apratim87/housingdata/data) (Kaggle) dataset,
-and 3) regression, with a custom estimator, using [Natality Baby Weight](https://catalog.data.gov/dataset?tags=birth-weight) (data.gov) dataset.
+1. **[template](https://github.com/GoogleCloudPlatform/cloudml-samples/tree/master/cloudml-template/template)**: The directory includes:
```
Comment (thedriftofwords) [inline cloudml-template/README.md]: Thanks for making this README look more structured! Throughout this section, please remove the commas / "ands" from the end of each list item. For example:

1) item 1,
2) item 2, and
3) item 3

becomes:

1) item 1
2) item 2
3) item 3

**Comment by ksalama** [comment]
@thedriftofwords - Thanks for your review. I have made the changes based on your recommendations.

**Comment by thedriftofwords** [comment]
Oops, I think one typo fix was missed. "Cloude MLE" -> "Cloud ML Engine" 
Otherwise, README looks good. Thanks!

**Comment by ksalama** [comment]
@thedriftofwords - Done :-)

**Comment by ksalama** [comment]
@dizcology : I added few tweaks to the config.yaml files:

1. example of a **discrete** param.
2. use **early stopping** option.
3. explicitly stating the **metric name** to optimise

File: cloudml-template/examples/german-custom/trainer/metadata.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+# task type can be either 'classification' or 'regression', based on the target feature in the dataset
+TASK_TYPE = 'custom'
+
+# list of all the columns (header) of the input data file(s)
+HEADER = ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'A10',
+          'A11', 'A12', 'A13', 'A14', 'A15', 'A16', 'A17', 'A18', 'A19', 'A20',
+          'class']
+
+# list of the default values of all the columns of the input data, to help decoding the data types of the columns
+HEADER_DEFAULTS = [['?'], [0.0], ['?'], ['?'], [0.0], ['?'], ['?'], [0.0], ['?'], ['?'],
```
Comment (dizcology) [inline cloudml-template/examples/german-custom/trainer/metadata.py]: is the usage of `'?'` here conventional?  (assuming it means there is no default?)

File: cloudml-template/template/trainer/model.py [hunk 5]
```diff
@@ -26,6 +26,35 @@
...
+    """
+    metrics = {}
+
+    # Example of implementing Root Mean Squared Error for regression
+
+    # pred_values = predictions['predictions']
+    # metrics['rmse'] = tf.metrics.root_mean_squared_error(labels=labels,
+    #                                                      predictions=pred_values)
+
+    # Example of implementing Sensitivity at Specificity for classification
```
Comment (dizcology) [inline cloudml-template/template/trainer/model.py]: The commented out code below seems to be for `mean_per_class_accuracy`.

File: cloudml-template/template/trainer/model.py [hunk 6]
```diff
@@ -123,58 +159,68 @@ def create_estimator(config):
...
+    estimator = tf.contrib.estimator.add_metrics(estimator, metric_fn)
 
-    return tf.estimator.Estimator(model_fn=_model_fn,
-                                  config=config)
+    return estimator
 
 
 # ***************************************************************************************
-# THIS IS A HELPER FUNCTION TO CREATE GET THE HIDDEN LAYER UNITS
+# HELPER FUNCTIONS USED FOR CONSTRUCTING THE MODELS
```
Comment (dizcology) [inline cloudml-template/template/trainer/model.py]: Should the user change these helper functions to adapt the template for their specific use case?

File: cloudml-template/examples/german-custom/setup.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+REQUIRED_PACKAGES = ['tensorflow>=1.4']
+
+setup(
+    name='trainer',
+    version='0.1',
+    install_requires=REQUIRED_PACKAGES,
+    packages=find_packages(),
+    include_package_data=True,
+    description='babyweight trainer package.'
```
[the remark this answers is not in the document]
Comment (ksalama) [reply cloudml-template/examples/german-custom/setup.py]: OK, the description is now changed to 'german trainer package'

File: cloudml-template/examples/german-custom/trainer/featurizer.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+    if categorical_columns_with_hash_bucket is not None:
+        feature_columns.update(categorical_columns_with_hash_bucket)
+
+    # add extended feature_column(s) before returning the complete feature_column dictionary
+    return extend_feature_columns(feature_columns)
+
+
+# **************************************************************************
+# YOU MAY NOT CHANGE THIS FUNCTION TO DEFINE WIDE AND DEEP COLUMNS
```
[the remark this answers is not in the document]
Comment (ksalama) [reply cloudml-template/examples/german-custom/trainer/featurizer.py]: OK, it is now changed to: YOU NEED NOT TO CHANGE THIS FUNCTION TO DEFINE WIDE AND DEEP COLUMNS

File: cloudml-template/examples/german-custom/trainer/metadata.py (hunk 4 above)
Comment (ksalama) [reply cloudml-template/examples/german-custom/trainer/metadata.py]: It doesn't matter which string value is used; can be 'NA', 'NULL', '?', or anything else...
It is just to till the tf csv decoder that this feature is a string.

Do you recommend specific value you use?

File: cloudml-template/template/trainer/model.py (hunk 5 above)
Comment (ksalama) [reply cloudml-template/template/trainer/model.py]: OK, the commented out code is changed to:

 # Example of implementing Mean per Class Accuracy for classification

File: cloudml-template/template/trainer/model.py (hunk 6 above)
Comment (ksalama) [reply cloudml-template/template/trainer/model.py]: OK, the comment is now changed to:

YOU NEED NOT TO CHANGE THESE HELPER FUNCTIONS USED FOR CONSTRUCTING THE MODELS

File: cloudml-template/README.md (hunk 3 above)
Comment (ksalama) [reply cloudml-template/README.md]: done

File: cloudml-template/examples/babyweight-custom/trainer/task.py [hunk 9]
```diff
@@ -146,6 +144,18 @@ def initialise_hyper_params(args_parser):
...
         type=float
     )
+    args_parser.add_argument(
+        '--learning-rate-decay-factor',
+        help="""\
+             **VALID FOR CUSTOM MODELS**
+             The factor by which the learning rate should decay by the end of the training.
+             decayed_learning_rate = learning_rate * decay_rate ^ (global_step / decay_steps)
+             If set to 1.0 (default), then no decay will occur
+             If set to 0.5, then the learning rate should reach 0.5 of its original value  \
```
[the remark this answers is not in the document]
Comment (ksalama) [reply cloudml-template/examples/babyweight-custom/trainer/task.py]: OK, given this formula (defined in the [docs](https://www.tensorflow.org/versions/r1.4/api_docs/python/tf/train/exponential_decay)): 

decayed_learning_rate = learning_rate * decay_rate ^ (global_step / decay_steps)

if the decay_rate=1, then learning_rate will still be the same, regardless of the value of the global_step or decay step, since 1 the power anything=1


if the decay_rate=0.5, then the learning rate will **reach** 0.5 of its original value at the end of the training. Please not that, global_step will always be **less than or equals** to decay_steps, since the decay_steps is set to train_steps

- in the update_learning_rate function in model.py, I have set:

    decay_steps = task.HYPER_PARAMS.train_steps  # decay after each training step

instead of:

    decay_steps = task.HYPER_PARAMS.num_epochs  # decay after each epoch

- I have also improved the description of the parameter in the task.py. as follows:

"If set to 0.5, then the learning rate should reach 0.5 of its original value at the end of the training. Note that, the decay_steps is set to train_steps"

File: cloudml-template/examples/babyweight-custom/trainer/model.py [hunk 10]
```diff
@@ -234,3 +212,24 @@ def construct_hidden_units():
...
+    Returns:
+       float - updated (decayed) learning rate
+    """
+    initial_learning_rate = task.HYPER_PARAMS.learning_rate
+    decay_steps = task.HYPER_PARAMS.num_epochs  # decay after each epoch
+    decay_factor = task.HYPER_PARAMS.learning_rate_decay_factor  # if set to 1, then no decay.
+
+    global_step = tf.train.get_global_step()
+
+    # decayed_learning_rate = learning_rate * decay_rate ^ (global_step / decay_steps)
```
[the remark this answers is not in the document]
Comment (ksalama) [reply cloudml-template/examples/babyweight-custom/trainer/model.py]: I updated decay_steps to be:

   decay_steps = task.HYPER_PARAMS.train_steps

Rather than:

  decay_steps = task.HYPER_PARAMS.num_epochs

As discussed in [docs](https://www.tensorflow.org/versions/r1.4/api_docs/python/tf/train/exponential_decay), using this formula:

decayed_learning_rate = learning_rate * decay_rate ^ (global_step / decay_steps)


The original learning_rate can only reach (learning_rate * decay_rate), when global_steps reaches decay_steps.

Which means, if learning_rate= 0.1, and decay_rate = 0.95, then at the end, learning_rate will be 0.1 * 0.95 = 0.095, and it cannot go below this value.

[the remark this answers is not in the document]
Comment (dizcology) [reply cloudml-template/examples/babyweight-custom/trainer/model.py]: Thanks - I misunderstood what `decay_steps` does.  The name suggests "every how many step will the learning rate be decayed".

File: cloudml-template/examples/german-custom/trainer/metadata.py (hunk 4 above)
Comment (dizcology) [reply cloudml-template/examples/german-custom/trainer/metadata.py]: Are the default values somehow used to fill in missing values, or only for specifying the type of the column?  I couldn't quite tell from the [docs](https://www.tensorflow.org/api_docs/python/tf/decode_csv).

I would guess that the type of `0.0` is `tf.float32` - is this correct?

Comment (ksalama) [reply cloudml-template/examples/german-custom/trainer/metadata.py]: Yes you are right, you use `0.0` for `tf.float32`. You can check the census example (also in this cloudml-sample repo [here](https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/census/estimator/trainer/model.py#L29) line 29.

File: cloudml-template/examples/babyweight-custom/trainer/model.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ksalama) [reply cloudml-template/examples/babyweight-custom/trainer/model.py]: I would call it `train_steps`, but the parameter is named `decay_steps` in the `tf.train.exponential_decay` [function](https://www.tensorflow.org/api_docs/python/tf/train/exponential_decay)

[the remark this answers is not in the document]
Comment (dizcology) [reply cloudml-template/examples/babyweight-custom/trainer/model.py]: It seems that the reason is `tf.train.exponential_decay` is used differently than the way you do here.  The example in the doc decays the learning rate with a factor of 0.96 for every 100K steps, while the training itself could take, say, 10M steps, with a final learning rate of about `0.016*initial_learning_rate`.  

My understanding is that is the common way of using exponential decay, and the reader come have expected.  Since you are recommending a different approach please consider adding to the comment "If set to 0.5, then..." mentioning this difference.

File: cloudml-template/examples/german-custom/trainer/metadata.py (hunk 4 above)
Comment (dizcology) [reply cloudml-template/examples/german-custom/trainer/metadata.py]: If it is indeed the case that these "default" values serve only the purpose of communicating types, please consider using the empty string for the string type.  `'?'` appears to be some magical convention.

**Comment by thedriftofwords** [comment]
There was a file conflict on README.md that I resolved on my end - mainly, changing the links within the repo to relative links.
